File: Packets/packets/BaseLocation.cs
using Game.Enums;
using System;
using System.Text;


public class BaseLocation : Packet<PacketType>
{
    public byte TeamIndex { get; set; }
    public double Longitude { get; set; }
    public double Latitude { get; set; }
    public BaseLocation(byte[] bytes) : base(bytes) { }

    public BaseLocation(Team team, GPS gps, PacketAction action) : base(PacketType.BaseLocation, action)
    {
        this.TeamIndex = (byte)team;
        this.Longitude = gps.Longitude;
        this.Latitude = gps.Latitude;
    }

    protected override void ReadPayload(byte[] bytes, byte offset)
    {
        TeamIndex = bytes[offset];
        Longitude = BitConverter.ToDouble(bytes, offset + 1);
        Latitude = BitConverter.ToDouble(bytes, offset + 9);
    }

    protected override void WritePayload(byte[] bytes, byte offset)
    {
        bytes[offset + 0] = TeamIndex;
        Array.Copy(BitConverter.GetBytes(Longitude), 0, bytes, offset + 1, 8);
        Array.Copy(BitConverter.GetBytes(Latitude), 0, bytes, offset + 9, 8);
    }

    public override string ToString()
    {
        return $"PlayerStatus{{teamIndex={TeamIndex}, longitude={Longitude}, latitude={Latitude}}}";
    }
}

File: src/main/java/net/nerfatg/proxy/packet/packets/BaseLocation.java
package net.nerfatg.proxy.packet.packets;

import net.nerfatg.Utils.GPS;
import net.nerfatg.proxy.packet.Packet;
import net.nerfatg.proxy.packet.PacketAction;
import net.nerfatg.proxy.packet.PacketType;
import net.nerfatg.Utils.GPS;

import java.nio.BufferOverflowException;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;

public class BaseLocation extends Packet {

    private String playerId;
    private byte teamIndex;
    private double longitude;
    private double latitude;

    public BaseLocation(ByteBuffer buffer) throws BufferOverflowException {
        super(buffer);
    }

    public BaseLocation(String playerId, byte teamIndex, GPS gps, PacketAction action) {
        super(PacketType.BaseLocation, action);

        this.playerId = playerId;
        this.teamIndex = teamIndex;
        this.longitude = gps.getLongitude();
        this.latitude = gps.getLatitude();
    }

    @Override
    public void readPayload(ByteBuffer buffer, int offset) throws BufferOverflowException {
        buffer.position(offset);
        buffer.order(ByteOrder.LITTLE_ENDIAN);

        byte[] playerIdBytes = new byte[8];
        buffer.get(playerIdBytes);
        this.playerId = new String(playerIdBytes, java.nio.charset.StandardCharsets.UTF_8);

        this.teamIndex = buffer.get();
        this.longitude = buffer.getDouble();
        this.latitude = buffer.getDouble();
    }

    @Override
    public void writePayload(ByteBuffer buffer, int offset) throws BufferOverflowException {
        buffer.position(offset);
        buffer.order(ByteOrder.LITTLE_ENDIAN);

        // Write playerId (8 bytes fixed)
        byte[] originalPlayerIdBytes = playerId.getBytes(java.nio.charset.StandardCharsets.UTF_8);
        buffer.put(originalPlayerIdBytes, 0, Math.min(originalPlayerIdBytes.length, 8));
        if (originalPlayerIdBytes.length < 8) {
            buffer.put(new byte[8 - originalPlayerIdBytes.length]); // pad with zeros
        }

        buffer.put(teamIndex);
        buffer.putDouble(longitude);
        buffer.putDouble(latitude);
    }

    public String getPlayerId() {return playerId; }
    public byte getTeamIndex() { return teamIndex; }
    public double getLongitude() { return longitude; }
    public double getLatitude() { return latitude; }

    @Override
    public String toString() {
        return String.format("BaseLocation{teamIndex=%d, longitude=%f, latitude=%f}", 
                           teamIndex, longitude, latitude);
    }
}
